File: dodam-application/dodam-rest-api/src/main/java/b1nd/dodam/restapi/bus/presentation/BusController.java
package b1nd.dodam.restapi.bus.presentation;

import b1nd.dodam.restapi.bus.application.BusUseCase;
import b1nd.dodam.restapi.bus.application.data.req.BusApplicantReq;
import b1nd.dodam.restapi.bus.application.data.req.BusBoardReq;
import b1nd.dodam.restapi.bus.application.data.req.BusReq;
import b1nd.dodam.restapi.bus.application.data.res.BusApplicantRes;
import b1nd.dodam.restapi.bus.application.data.res.BusDetailRes;
import b1nd.dodam.restapi.bus.application.data.res.BusRes;
import b1nd.dodam.restapi.support.data.Response;
import b1nd.dodam.restapi.support.data.ResponseData;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/bus")
@RequiredArgsConstructor
class BusController {
    private final BusUseCase useCase;

    @GetMapping
    public ResponseData<List<BusRes>> buses() {
        return useCase.getBus();
    }

    @GetMapping("/{id}/seats")
    public ResponseData<List<Integer>> requestedSeats(@PathVariable long id) {
        return useCase.getRequestedSeats(id);
    }

    @GetMapping("/{id}")
    public ResponseData<BusDetailRes> busDetail(@PathVariable long id) {
        return useCase.busDetail(id);
    }

    @GetMapping("/my")
    public ResponseData<BusApplicantRes> getMy() {
        return useCase.my();
    }

    @PostMapping
    public Response createBus(@RequestBody BusReq req) {
        return useCase.createBus(req);
    }

    @PostMapping("/board")
    public Response createBusApplicationStudent(@RequestBody BusApplicantReq req) {
        return useCase.createBusApplicant(req);
    }

    @PostMapping("/board/{seat}")
    public Response applyBus(@PathVariable int seat) {
        return useCase.apply(seat);
    }

    @PutMapping("/{id}")
    public Response modifyBus(@PathVariable long id, @RequestBody BusReq req) {
        return useCase.updateBus(id, req);
    }

    @PatchMapping("board/{seat}")
    public Response changeSeat(@PathVariable int seat) {
        return useCase.changeSeat(seat);
    }

    @PatchMapping("/board")
    public Response changeBoard(@RequestBody BusBoardReq req) {
        return useCase.changeBoardingType(req.boardingType());
    }

    @DeleteMapping("/{id}")
    public Response deleteBus(@PathVariable long id) {
        return useCase.deleteBus(id);
    }
}

File: services/service-bus/src/main/kotlin/com/b1nd/dodamdodam/bus/presentation/bus/BusController.kt
package com.b1nd.dodamdodam.bus.presentation.bus

import com.b1nd.dodamdodam.bus.application.bus.BusUseCase
import com.b1nd.dodamdodam.bus.application.bus.data.request.BusApplicantRequest
import com.b1nd.dodamdodam.bus.application.bus.data.request.BusBoardRequest
import com.b1nd.dodamdodam.bus.application.bus.data.request.BusRequest
import com.b1nd.dodamdodam.bus.application.bus.data.response.BusApplicantResponse
import com.b1nd.dodamdodam.bus.application.bus.data.response.BusDetailResponse
import com.b1nd.dodamdodam.bus.application.bus.data.response.BusResponse
import com.b1nd.dodamdodam.core.common.data.Response
import com.b1nd.dodamdodam.core.security.annotation.authentication.UserAccess
import com.b1nd.dodamdodam.core.security.passport.enumerations.RoleType
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class BusController(
    private val busUseCase: BusUseCase
) {
    @GetMapping("/bus")
    fun getBuses(): Response<List<BusResponse>> {
        return busUseCase.getBuses()
    }

    @GetMapping("/bus/{id}/seats")
    fun getRequestedSeats(@PathVariable id: Long): Response<List<Int?>> {
        return busUseCase.getRequestedSeats(id)
    }

    @GetMapping("/bus/{id}")
    fun busDetail(@PathVariable id: Long): Response<BusDetailResponse> {
        return busUseCase.busDetail(id)
    }

    @UserAccess
    @GetMapping("/bus/my")
    fun getMy(): Response<BusApplicantResponse?> {
        return busUseCase.my()
    }

    @UserAccess(roles = [RoleType.TEACHER, RoleType.ADMIN])
    @PostMapping("/bus")
    fun createBus(@RequestBody request: BusRequest): Response<Unit> {
        return busUseCase.createBus(request)
    }

    @UserAccess(roles = [RoleType.TEACHER, RoleType.ADMIN])
    @PostMapping("/bus/board")
    fun createBusApplicant(@RequestBody request: BusApplicantRequest): Response<Unit> {
        return busUseCase.createBusApplicant(request)
    }

    @UserAccess
    @PostMapping("/bus/board/{seat}")
    fun apply(@PathVariable seat: Int): Response<Unit> {
        return busUseCase.apply(seat)
    }

    @UserAccess(roles = [RoleType.TEACHER, RoleType.ADMIN])
    @PutMapping("/bus/{id}")
    fun updateBus(@PathVariable id: Long, @RequestBody request: BusRequest): Response<Unit> {
        return busUseCase.updateBus(id, request)
    }

    @UserAccess
    @PatchMapping("/bus/board/{seat}")
    fun changeSeat(@PathVariable seat: Int): Response<Unit> {
        return busUseCase.changeSeat(seat)
    }

    @UserAccess
    @PatchMapping("/bus/board")
    fun changeBoard(@RequestBody request: BusBoardRequest): Response<Unit> {
        return busUseCase.changeBoardingType(request)
    }

    @UserAccess(roles = [RoleType.TEACHER, RoleType.ADMIN])
    @DeleteMapping("/bus/{id}")
    fun deleteBus(@PathVariable id: Long): Response<Unit> {
        return busUseCase.deleteBus(id)
    }
}
